runtime/platform/test/executor_pal_runtime_override_test.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gtest/gtest.h>
+
+#include <executorch/runtime/platform/platform.h>
+#include <executorch/runtime/platform/test/pal_spy.h>
+
+namespace {
+PalSpy* active_spy;
+
+void pal_init(void) {
+  active_spy->init();
+}
+
+et_timestamp_t pal_current_ticks(void) {
+  return active_spy->current_ticks();
+}
+
+et_tick_ratio_t pal_ticks_to_ns_multiplier(void) {
+  return active_spy->ticks_to_ns_multiplier();
+}
+
+void pal_emit_log_message(
+    et_timestamp_t timestamp,
+    et_pal_log_level_t level,
+    const char* filename,
+    ET_UNUSED const char* function,
+    size_t line,
+    const char* message,
+    ET_UNUSED size_t length) {
+  active_spy->emit_log_message(
+      timestamp, level, filename, function, line, message, length);
+}
+
+void* pal_allocate(size_t size) {
+  return active_spy->allocate(size);
+}
+
+void pal_free(void* ptr) {
+  active_spy->free(ptr);
+}
+} // namespace
+
+class RuntimePalOverrideTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    // Capture the current PAL implementation so that it can be restored
+    // after the test.
+    _original_pal_impl = *executorch::runtime::get_pal_impl();
+  }
+
+  void TearDown() override {
+    // Restore the original PAL implementation.
+
+    // This is a slightly hacky way to allow this test to exist alongside
+    // the executor_pal_override_test, which provides a build-time override
+    // for et_pal_init. This implementation asserts that an intercept exists.
+    // Since register_pal calls init, we need to make sure that an intercept
+    // is registered. It will be deregistered when it goes out of scope,
+    // allowing the tests to run in any order.
+    InterceptWith iw(_spy);
+    auto success = executorch::runtime::register_pal(_original_pal_impl);
+    if (!success) {
+      throw std::runtime_error("Failed to restore PAL implementation.");
+    }
+  }
+
+  void RegisterSpy() {
+    active_spy = &_spy;
+
+    executorch::runtime::register_pal(executorch::runtime::pal_table(
+        pal_init,
+        nullptr, // abort
+        pal_current_ticks,
+        pal_ticks_to_ns_multiplier,
+        pal_emit_log_message,
+        pal_allocate,
+        pal_free,
+        __FILE__));
+  }
+
+  PalSpy _spy;
```

Comment:
what's spy?

Reply:
Seems like a test implementation

Reply:
Yeah, this is an existing test class that's used in the other PAL tests. I reused it here for simplicity.

Reply:
No I am wondering about the name lol

Reply:
I assume it's called Spy as it's implementing a spy pattern for testing - where it observes (counts, in this case) method calls made on it. Not sure who originally added it.